LibUnicode: Implement Unicode Normalization

matcool opened the pull request:
This is WIP as there are a couple of things missing, and code could be improved, so any help would be nice :-)

---

This adds support for normalization forms D and C (see https://unicode.org/reports/tr15/), as compatibility mappings require more data to be stored and i'm not sure on how to store it..


Missing:
- [ ] NFKD, NFKC support
- [ ] Hangul support
- [ ] Quick Check?
- [ ] Better way of searching a code point from its mappings

trflynn89 reviewed (commented):
Only had time to review the fist commit, just a couple minor comments on the code itself.

Looking ahead, it looks like code points may have a decomposition mapping of more than 2 code points? e.g. U+FDFA maps to `U+0635 U+0644 U+0649 U+0020 U+0627 U+0644 U+0644 U+0647 U+0020 U+0639 U+0644 U+064A U+0647 U+0020 U+0648 U+0633 U+0644 U+0645` - 18 code points! :scream: 

That'll of course bash heads with the `Array<u32, 2>` storing the mappings here.

What you can do instead is change the `Array` to a `Vector<u32>` (only in the parser), and just append each mapped code point to that vector.

Then when you generate the `CodePointDecomposition` structures, you can first generate each code point's decomposition mapping as standalone `Array`s, and then store a `Span` of those code points in the `CodePointDecomposition` structure.

So if we look at just the code points U+00B9, U+00BA, U+00BC, U+00BD, U+00BE, U+00C0, this might look like:

```c++
static constexpr Array<u32, 1> s_b9_decomposition { 0x31 };
static constexpr Array<u32, 1> s_ba_decomposition { 0x6F };
static constexpr Array<u32, 3> s_bc_decomposition { 0x31, 0x2044, 0x34 };
static constexpr Array<u32, 3> s_bd_decomposition { 0x31, 0x2044, 0x32 };
static constexpr Array<u32, 3> s_be_decomposition { 0x33, 0x2044, 0x34 };
static constexpr Array<u32, 2> s_c0_decomposition { 0x41, 0x300 };

struct CodePointDecomposition {
    u32 code_point { 0 };
    bool canonical { true };
    Span<u32 const> decomposition;
};

static constexpr Array<CodePointDecomposition, 6> s_decomposition_mappings { {
    { 0xb9, false, s_b9_decomposition.span() },
    { 0xba, false, s_ba_decomposition.span() },
    { 0xbc, false, s_bc_decomposition.span() },
    { 0xbd, false, s_bd_decomposition.span() },
    { 0xbe, false, s_be_decomposition.span() },
    { 0xc0, true, s_c0_decomposition.span() },
} };
```

matcool commented:
> mapping of more than 2 code points? 
Yup, those are the compatibility mappings I've been avoiding :)

Thanks for the span idea, much better than making a vector for each mapping

trflynn89 commented on Meta/Lagom/Tools/CodeGenerators/LibUnicode/GenerateUnicodeData.cpp:
```diff
@@ -732,8 +774,36 @@ struct SpecialCasing {
+
 Optional<Locale> locale_from_string(StringView locale);
 
+Span<CodePointDecomposition const> code_point_decompositions();
```
Which, since the next commit adds Normalize.cpp anyways, it'd be a bit better if both `code_point_decomposition` and `code_point_decompositions` were declared in Normalize.h (with weak definitions in Normalize.cpp) to be closer to where it's actually used.

matcool replied:
Done in 784beff :+1:

trflynn89 commented on Meta/Lagom/Tools/CodeGenerators/LibUnicode/GenerateUnicodeData.cpp:
```diff
@@ -1094,6 +1191,18 @@ Optional<StringView> code_point_abbreviation(u32 code_point)
+    return *mapping;
+}
+
+Span<CodePointDecomposition const> code_point_decompositions() {
```
Minor, code style is to put the opening `{` of functions on a new line:

```c++
Span<CodePointDecomposition const> code_point_decompositions()
{
```

trflynn89 commented on Userland/Libraries/LibUnicode/Forward.h:
```diff
@@ -22,5 +22,6 @@ enum class WordBreakProperty : u8;
 struct CurrencyCode;
 struct Emoji;
 struct SpecialCasing;
+struct CodePointDecomposition;
```
Let's keep this list alphabetical

trflynn89 commented on Userland/Libraries/LibUnicode/Normalize.cpp:
```diff
@@ -0,0 +1,196 @@
+            // "When a canonical mapping consists of a pair of characters,
+            //  the first character may itself be a character with a decomposition mapping,
+            //  but the second character never has a decomposition mapping."
+            if (mapping->decomposition.size() > 1 && mapping->decomposition[1] != 0) {
```
It doesn't look like the second decomposed code point is ever 0 in the UCD? i.e. I think you can drop the check for `mapping->decomposition[1] != 0` (unless I'm missing something)

matcool replied:
Right, this is a left over from when the decomposition was stored in an Array<u32, 2>, so 0 was used as an empty value

trflynn89 commented on Userland/Libraries/LibUnicode/Normalize.cpp:
```diff
@@ -0,0 +1,196 @@
+// This can be any sorting algorithm that maintains order (like std::stable_sort),
+// however bubble sort is the easier to implement, so go with it.
+template<typename T, typename LessThan>
+void bubble_sort(Span<T> span, LessThan less_than)
```
Hmm we should make the single pivot quick sort stable if we can (not a comment for you, just talking out loud)

trflynn89 commented on Userland/Libraries/LibUnicode/Normalize.cpp:
```diff
@@ -0,0 +1,196 @@
+}
+
+// The Canonical Ordering Algorithm, as specified in Version 15.0.0 of the Unicode Standard.
+// See Section 3.11, D109; and UAX #15 https://unicode.org/reports/tr15
```
Let's add direct links to the spec (the PDF is kinda annoying to find on unicode.org):
https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G49591

matcool replied:
Sure, although just curious, how did you get `#G49591` from the D109 point?

trflynn89 commented on Userland/Libraries/LibUnicode/Normalize.cpp:
```diff
@@ -0,0 +1,196 @@
+}
+
+// The Canonical Composition Algorithm, as specified in Version 15.0.0 of the Unicode Standard.
+// See Section 3.11, D117; and UAX #15 https://unicode.org/reports/tr15
```
https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G50628

trflynn89 commented on Userland/Libraries/LibUnicode/Normalize.cpp:
```diff
@@ -0,0 +1,196 @@
+}
+
+// See Section 3.11, D115 of Version 15.0.0 of the Unicode Standard.
+static bool is_blocked(Span<u32> span, size_t a, size_t c)
```
`span` isn't a good variable name, let's call it `code_points` or something similar

trflynn89 commented on Userland/Libraries/LibUnicode/Normalize.cpp:
```diff
@@ -0,0 +1,196 @@
+
+// The Canonical Composition Algorithm, as specified in Version 15.0.0 of the Unicode Standard.
+// See Section 3.11, D117; and UAX #15 https://unicode.org/reports/tr15
+static void canonical_composition_algorithm(Vector<u32>& input)
```
Same comment on `input` as `span` - because it's also serving as an output, let's call it `code_points` as well.

trflynn89 commented on Meta/Lagom/Tools/CodeGenerators/LibUnicode/GenerateUnicodeData.cpp:
```diff
@@ -809,6 +809,7 @@ static ErrorOr<void> generate_unicode_data_implementation(Core::Stream::Buffered
 #include <AK/StringView.h>
 #include <LibUnicode/CharacterTypes.h>
 #include <LibUnicode/UnicodeData.h>
+#include <LibUnicode/Normalize.h>
```
This commit should be squashed into the first commit. It's considered a fixup, which we actively avoid within a PR (see [here](https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy)).

trflynn89 replied:
Right, so you would add Normalize.h and Normalize.cpp in the first commit as well.

matcool replied:
Both now merged into [c8c4aaa](https://github.com/SerenityOS/serenity/pull/15464/commits/c8c4aaafdadafc4b1fdac80a3d469073f09b3686)

trflynn89 reviewed (commented):
Generator looks good now! Some comments on the normalization algorithms in the second commit - I'm still digesting it a bit so bear with me :laughing:

matcool commented on Userland/Libraries/LibUnicode/Normalize.h:
```diff
@@ -0,0 +1,32 @@
+    NFKC
+};
+
+[[nodiscard]] Vector<u32> normalize(NormalizationForm form, Utf32View string);
```
UTF-32 makes it easier to just iterate through the code points, which is what most of the algorithms do. Would having a StringView overload that then converts it to UTF-32 be fine?

matcool replied:
Nevermind, just a public StringView overload is fine.

trflynn89 replied:
:+1: In case you haven't found it, you can take a `StringView` as a parameter, and wrap it in a `Utf8View` inside your function. That will let you iterate over the same way you're doing now, one decoded code point at a time.

matcool commented:
> Hm, would this be a better api?
> 
> ```c++
> [[nodiscard]] String normalize(StringView string, NormalizationForm form = NormalizationForm::NFC);
> ```

Switched to this order, and also added support for hangul (de)composition, meaning there's nothing required left :-)

trflynn89 reviewed (approved):
Looks good to me! Thanks for implementing this :rocket: 

Just curious, what are you planning on using normalization for next?
